Rollup-based build recipe

jeffposnick opened the pull request:
R: @jpmedley @addyosmani @gauntface

Here's a recipe for how to structure your service worker using ES2015 imports, and then bundle that source file into a final service worker using `rollup`. It has examples for kicking off that process as both part of an `npm scripts` chain or as `gulp` tasks.

I don't know whether I need to do anything more than drop this `.md` file into a directory, so please let me know if there are other things I should do to make sure this gets picked up.

gauntface commented on docs/recipes/build/rollup.md:
```diff
@@ -0,0 +1,135 @@
+# [Rollup](https://github.com/rollup/rollup) Bundling
+
+## Dependencies
```
There needs to be a title and intro that explains what this doc is for / going to achieve.

gauntface commented on docs/recipes/build/rollup.md:
```diff
@@ -0,0 +1,135 @@
+
+## Service Worker Code
+
+This approach assumes that your `src/sw.js` file is using [ES2015 module
```
This could be part of the intro - i.e. explanation of if you are using Rollup and would like to use ES2015 modules in your service worker - here is how you can get the manifest revision details.

gauntface commented on docs/recipes/build/rollup.md:
```diff
@@ -0,0 +1,135 @@
+
+## Option 1: Vanilla node Build Script
+
+The following is a basic `node` build script that can be run as part of a larger
```
I don't understand this paragraph - what is the goal?

Why is it referencing npm scripts?

gauntface commented on docs/recipes/build/rollup.md:
```diff
@@ -0,0 +1,135 @@
+The following is a basic `node` build script that can be run as part of a larger
+build process, e.g. in `package.json` at the end of an
+[`npm scripts`](https://docs.npmjs.com/misc/scripts) `"build"` chain as
+`&& node build-sw.js`.
```
Is it explained after this section?

jpmedley commented on docs/recipes/build/rollup.md:
```diff
@@ -0,0 +1,135 @@
+
+## Dependencies
+
+- A working installation of `node` (4.0+) & `npm`.
```
'A working installation of' to 'Working installations of'.

gauntface commented on docs/recipes/build/rollup.md:
```diff
@@ -0,0 +1,135 @@
+// manifest generation ->
+//   rollup configuration ->
+//   write bundle to disk
+swBuild.generateFileManifest(swBuildConfig)
```
I'd inline config here - It's weird go sw-build config, rollup config, sw-build, then rollup promise.

gauntface commented on docs/recipes/build/rollup.md:
````diff
@@ -0,0 +1,135 @@
+
+```
+
+## Option 2: [gulp](http://gulpjs.com/) Tasks
````
Maybe refrain from using the term "Options" and keep to recipes - i.e. using sw-build with Rollup and Gulp

gauntface commented on docs/recipes/build/rollup.md:
```diff
@@ -0,0 +1,135 @@
+The same basic code can be split up into two separate `gulp` tasks, if you
+already are using that for your build process.
+
+**Note:** As with before, make sure that you run the `write-sw` task at the very
```
Maybe bring this out in a note in the intro and why it's important and then no need to mention again.

gauntface commented on docs/recipes/build/rollup.md:
```diff
@@ -0,0 +1,135 @@
+  });
+});
+
+gulp.task('write-sw', ['write-manifest'], () => {
```
real nit - but would bundle-sw.js be a better name? I'm just thinking that write-manifest is generating out of thin air and sw.js will be taking an existing file and spitting it out the other side.

jeffposnick commented:
Thanks for the feedback, and PTAL.

gauntface commented on .gitignore:
```diff
@@ -4,3 +4,4 @@ temp
 npm-debug.log
 packages/**/LICENSE
 .DS_Store
+!docs/recipes/build
```
Where is this from? Just curious - not blocking.

jeffposnick replied:
We've got an entry a few lines up that ignores `build` (in the context of the build output for a given module).

This is just to negate that so that I'm able to call the directory that hosts the recipes `build` and still check it in to git.

gauntface commented on docs/recipes/build/rollup.md:
```diff
@@ -0,0 +1,144 @@
+- You are comfortable using [`Rollup`](https://github.com/rollup/rollup) to
+handle transforming the `sw.js` file you maintain into a production-ready,
+self contained service worker file, containing the `sw-lib` library code along
+with your custom logic.
```
I'm not going to block because this going to be personal, but I kind of prefer the second half of this intro and think you could go along the lines of:

"If you are using Rollup and ES2015 imports, you can use sw-lib as an import in your service worker and sw-build can output a precache manifest that can be imported as well.

This recipe illustrates how you can use ES2015 imports in your service worker with sw-lib and sw-build."

jeffposnick replied:
I think I'll keep as-is. This will probably get tweaked more based on hackathon feedback.

gauntface commented on docs/recipes/build/rollup.md:
```diff
@@ -0,0 +1,144 @@
+
+## Your Build Process
+
+Here are two alternatives that lead to an equivalent, ready-to-deploy service
```
This feels like a wordy way to say:

"You have two options, you can build your service worker using Gulp or npm scripts"

jeffposnick replied:
There are more than two options, but I'll rephrase to make it less wordy 😄

gauntface commented on docs/recipes/build/rollup.md:
```diff
@@ -0,0 +1,144 @@
+
+### `npm scripts`
+
+The following is a basic `node` script that can be run as part of a larger build
```
Nit: I'd add an example `"scripts": {"build": "node ./build.js"}` as it's not clear where / what this Javascript / should do.

jeffposnick replied:
If you're following this recipe, it's not like to look like `"scripts": {"build": "node ./build.js"}`

More likely is `"scripts": {"build": "node build.js && node build-sw.js"}` (chaining the additional build process outlined here to the end of your existing build process).

But I can include that sample.

gauntface commented on docs/recipes/build/rollup.md:
````diff
@@ -0,0 +1,144 @@
+```js
+// Contents of src/sw.js:
+
+import manifest from '/tmp/manifest.js';
````
Is it worth highlighting after this code sample that in the npm run script / gulp process you'll write the manifest file using sw-build before running the service worker file through rollup?

jeffposnick replied:
I've added an inline comment to that effect.

jeffposnick commented:
I've addressed the feedback, hopefully in a satisfactory fashion.

Tweaking prose in a PR vs. in a Google Doc is always a pain, so this might not be the perfect final version, but I'm happy to address any changes in a follow-up PR.

GitHub won't let me merge until @jpmedley approves this, though.
